folly/io/async/AsyncSSLSocket.cpp

```diff
@@ -1359,6 +1397,63 @@ void AsyncSSLSocket::handleConnect() noexcept {
   AsyncSocket::handleInitialReadWrite();
 }
 
+#if FOLLY_OPENSSL_HAS_TLS13
+int AsyncSSLSocket::handleReadEarlyData() noexcept {
+  VLOG(3) << "AsyncSSLSocket::handleReadEarlyData() this=" << this
+          << ", fd=" << fd_ << ", state=" << int(state_) << ", "
+          << "sslState=" << sslState_ << ", events=" << eventFlags_;
+  size_t bufSize = SSL_CTX_get_max_early_data(ctx_->getSSLCtx());
+  unsigned char buf[bufSize];
+  size_t readBytes;
+  int edRet = SSL_READ_EARLY_DATA_ERROR;
+  int sslErr = SSL_ERROR_NONE;
+
+  edRet = SSL_read_early_data(ssl_.get(), buf, bufSize, &readBytes);
+  VLOG(3) << "AsyncSSLSocket::handleReadEarlyData()"
+          << " SSL_read_early_data return Code: "
+          << earlyDataErrCodeString(edRet)
+          << ", Early data OpenSSL Status: "
+          << earlyDataStatusString(ssl_.get());
+  switch(edRet) {
+    case SSL_READ_EARLY_DATA_ERROR:
+      sslErr = SSL_get_error(ssl_.get(), 0);
+      VLOG(3) << "SSL Error: " << sslErr;
+      if (SSL_ERROR_SSL == sslErr || SSL_ERROR_SYSCALL == sslErr) {
+        // Unrecoverable error in OpenSSL lib occured due to abrupt EOF sent
+        // by peer. Aborting the early data phase.
+        earlyDataPhaseComplete_ = true;
+        break;
+      }
+      // Either SSL_accept or SSL_read_ex inside SSL_read_early_data() returned
+      // SSL_ERROR_WANT_READ. Retry the operation by calling
+      // SSL_read_early_data again in next read event
+      updateEventRegistration(EventHandler::READ, EventHandler::WRITE);
+      sslState_ = STATE_ACCEPTING;
+      break;
+    case SSL_READ_EARLY_DATA_SUCCESS:
+      if (ctx_->parseEarlyDataCb_ &&
+          !ctx_->parseEarlyDataCb_(buf, readBytes)) {
+        VLOG(3) << "Early data content rejected ";
+        sslState_ = STATE_ERROR;
+        AsyncSocketException ex(AsyncSocketException::EARLY_DATA_REJECTED,
+                                "Invalid Early data content");
+        failHandshake(__func__, ex);
+        break;
+      }
```

Comment:
This does not mean that you fully received all early data application records. Quoting OpenSSL's documentation:

> SSL_READ_EARLY_DATA_SUCCESS
This indicates that early data was successfully read. This should be treated in the same way as a 1 return value from SSL_read_ex(3). **You should continue to call SSL_read_early_data() to read more data.**

(emphasis mine).